### PR TITLE
Update go lockfile generator to use output from `go list`

### PR DESCRIPTION
**Pull request by ejortega**
Changes the `go` lockfile generator to use the output of `go list -m -json all` from the previous `go.sum` generation using `go get -d`.

Closes https://github.com/phylum-dev/cli/issues/1396

## Checklist

- [x] Does this PR have an associated issue (i.e., `closes #<issueNum>` in description above)?
- [x] Have you ensured that you have met the expected acceptance criteria?
- [x] Have you created sufficient tests?
- [x] Have you updated all affected documentation?
- [x] Have you updated CHANGELOG.md (or extensions/CHANGELOG.md), if applicable

### CODE REVIEW

File: lockfile_generator/src/go.rs [hunk 1]
```diff
@@ -1,31 +1,101 @@
...
+            return Err(Error::NonZeroExit(code, stderr.into()));
+        }
+        // Parse go list STDOUT.
+        let stdout = String::from_utf8(output.stdout)?;
+        let stream = serde_json::Deserializer::from_str(&stdout).into_iter::<Module>();
+        let packages = stream.filter_map(|res| res.ok()).filter_map(create_go_dep);
+
+        let mut lockfile = String::new();
+        for pkg in packages {
+            let _ = writeln!(lockfile, "{} {} h1:h1", pkg.name, pkg.version);
```
Comment (kylewillmon) [inline lockfile_generator/src/go.rs:93]: It's probably a good idea to make it obvious that the second `h1` here is a fake hash value.

```suggestion
            let _ = writeln!(lockfile, "{} {} h1:FAKEHASH", pkg.name, pkg.version);
```

Also, [the hash is base64-encoded][1]. So as an added bonus, `FAKEHASH` is valid base64.

[1]: https://go.dev/ref/mod#go-sum-files

File: tests/fixtures/lock_generate_go/go.sum [hunk 2]
```diff
@@ -0,0 +1,17 @@
+github.com/coreos/go-systemd/v22 v22.5.0/go.mod h1:Y58oyj3AT4RCenI/lSvhwexgC+NSVTIJ3seZv2GcEnc=
```
Comment (kylewillmon) [inline tests/fixtures/lock_generate_go/go.sum:1]: This is a bit beyond the scope of this PR, but I'm curious about this...

This dependency (and a few others) only have hashes for `/go.mod` and not for the entire package. But this appears to be a legitimate dependency... How does `go` validate the rest of the package?

**Review by kylewillmon — CHANGES_REQUESTED** [review]

File: tests/fixtures/lock_generate_go/main.go
Comment (cd-work) [inline tests/fixtures/lock_generate_go/main.go:1]: Why are we adding an entire project here? Is this necessary to make the command work? If so, is it possible that will cause issues with our lockfile generator considering previously we only needed a single file, while now auxiliary files need to be present in the same directory?

The actual code part certainly isn't necessary, so it should probably be removed since it just distracts from the bits we care about.

File: tests/fixtures/lock_generate_go/go.sum (hunk 2 above)
Comment (ejortega) [reply tests/fixtures/lock_generate_go/go.sum:1]: Actually while answering @cd-work question, it appears the command `go list -m -json all` is parsing the `go.sum`. We might be able to just adjust the parser to produce the same results.

File: tests/fixtures/lock_generate_go/main.go
Comment (cd-work) [reply tests/fixtures/lock_generate_go/main.go:1]: > I was following the example for [csharp](https://github.com/phylum-dev/cli/tree/main/tests/fixtures/lock_generate_csproj), which also includes a project and even a .gitignore, which is probably unnecessary.

The csharp is a minimal generated project, your code doesn't look minimal to me?

Comment (ejortega) [reply tests/fixtures/lock_generate_go/main.go:1]: I created it solely to get a few dependencies in. I can make it more minimal

File: tests/fixtures/lock_generate_go/go.sum (hunk 2 above)
Comment (matt-phylum) [reply tests/fixtures/lock_generate_go/go.sum:1]: I thought this work started because in newer versions of go, `go list` was parsing the go.sum file and then recursing into go.sum files downloaded from the internet.

Comment (ejortega) [reply tests/fixtures/lock_generate_go/go.sum:1]: We use the `go.sum` as a lockfile, which is incorrect. It can lead to having false positives as includes packages not used to in a project.

Comment (cd-work) [reply tests/fixtures/lock_generate_go/go.sum:1]: > It can lead to having false positives as includes packages not used to in a project.

This is true for multiple ecosystems, is it not? Cargo will add things to the lockfile as long as they're in the manifest, it doesn't matter if it's used or not. I don't really see an issue with that.

Comment (ejortega) [reply tests/fixtures/lock_generate_go/go.sum:1]: I wasn't aware of that from cargo, but according to a customer, the output from `go list` was a 95% reduction in the amount of packages listed in the go.sum.

File: lockfile/src/golang.rs [hunk 3]
```diff
@@ -66,4 +66,69 @@ mod tests {
...
     }
+
+    #[cfg(feature = "generator")]
+    #[test]
+    fn lock_generate_goproj() {
+        let path = Path::new("../tests/fixtures/lock_generate_go/go.mod");
+        let lockfile = GoGenerator.generate_lockfile(path).unwrap();
+
+        let pkgs = GoSum.parse(&lockfile).unwrap();
+        assert_eq!(pkgs.len(), 9);
```
[the remark this answers is not in the document]
Comment (cd-work) [reply lockfile/src/golang.rs:77]: Has not really be an issue in the past. Figuring it out is usually very straight-forward and logging all packages is easy.

File: tests/fixtures/lock_generate_go/go.sum (hunk 2 above)
Comment (matt-phylum) [reply tests/fixtures/lock_generate_go/go.sum:1]: It does not work offline with these files:

```
root@bf8b095c5ee5:/app# go list -m -json all
go: github.com/go-chi/chi/v5@v5.0.12: Get "https://proxy.golang.org/github.com/go-chi/chi/v5/@v/v5.0.12.info": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
go: github.com/mattn/go-colorable@v0.1.13: Get "https://proxy.golang.org/github.com/mattn/go-colorable/@v/v0.1.13.info": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
go: github.com/mattn/go-isatty@v0.0.19: Get "https://proxy.golang.org/github.com/mattn/go-isatty/@v/v0.0.19.info": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
go: github.com/rs/zerolog@v1.32.0: Get "https://proxy.golang.org/github.com/rs/zerolog/@v/v1.32.0.info": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
go: golang.org/x/sys@v0.12.0: Get "https://proxy.golang.org/golang.org/x/sys/@v/v0.12.0.info": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
go: github.com/go-chi/chi/v5@v5.0.12: Get "https://proxy.golang.org/github.com/go-chi/chi/v5/@v/v5.0.12.mod": dial tcp: lookup proxy.golang.org on 192.168.65.7:53: dial udp 192.168.65.7:53: connect: network is unreachable
```

It tries to resolve the packages from go.mod, even if go.sum is available.

Comment (ejortega) [reply tests/fixtures/lock_generate_go/go.sum:1]: Ah I tried the same thing but it worked however had network up. With `go get -d` and `go tidy`, there's some messages to stdout that shows the files being downloaded, but not with `go list -m -json all`.

File: lockfile/src/golang.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cd-work) [reply lockfile/src/golang.rs:77]: Your sorting doesn't actually fix this here.

[the remark this answers is not in the document]
Comment (ejortega) [reply lockfile/src/golang.rs:77]: Then I'm not sure what the issue is.

[the remark this answers is not in the document]
Comment (matt-phylum) [reply lockfile/src/golang.rs:77]: When asserting that the collections are equal, don't compare the lengths. Alternatively, if you're doing tests for whether the collection contains the expected elements, comparing the length last would give more useful output in the most likely failure case where something is missing. When the length assertion fails it just says the length is wrong and you have to step through with a debugger or modify the code in order to know why the lengths are different.

[the remark this answers is not in the document]
Comment (ejortega) [reply lockfile/src/golang.rs:77]: Gotcha, I was just being consistent with [other tests](https://github.com/phylum-dev/cli/blob/c6e1f78deda3c02bef73d6a70252a8599430d01d/lockfile/src/csharp.rs#L166). No one seemed to have an issue there. There's a test further below that actually tests if they are equal, but I'll remove this length check.

[the remark this answers is not in the document]
Comment (cd-work) [reply lockfile/src/golang.rs]: > No one seemed to have an issue there.

As I've said, this has never been an issue. I don't mind it at all personally.

That said, most other tests do this because comparing the entire vec is just unnecessary noise, so they compare the length to check for approximate correctness and then take a few edgecase samples to confirm specific values.

File: tests/fixtures/lock_generate_go/go.sum (hunk 2 above)
Comment (kylewillmon) [reply tests/fixtures/lock_generate_go/go.sum:1]: This still didn't sit right with me... So I decided to test this by adding some source code to the `tests/fixtures/lock_generate_go`. Specifically, I added this file (as `main.go`):

```go
package example

import (
  "io"

  "github.com/rs/zerolog/journald"
)

func Foo() io.Writer {
  return journald.NewJournalDWriter()
}
```

And then running `go build` fails with the following error:

```sh-session
> go build
/Users/kyle/go/pkg/mod/github.com/rs/zerolog@v1.32.0/journald/journald.go:27:2: missing go.sum entry for module providing package github.com/coreos/go-systemd/v22/journal (imported by github.com/rs/zerolog/journald); to add:
	go get github.com/rs/zerolog/journald@v1.32.0
```

Running the requested command yields the following:

```sh-session
> go get github.com/rs/zerolog/journald@v1.32.0

> git diff
diff --git a/tests/fixtures/lock_generate_go/go.mod b/tests/fixtures/lock_generate_go/go.mod
index d73b228..5750659 100644
--- a/tests/fixtures/lock_generate_go/go.mod
+++ b/tests/fixtures/lock_generate_go/go.mod
@@ -8,6 +8,7 @@ require (
 )

 require (
+       github.com/coreos/go-systemd/v22 v22.5.0 // indirect
        github.com/mattn/go-colorable v0.1.13 // indirect
        github.com/mattn/go-isatty v0.0.19 // indirect
        golang.org/x/sys v0.12.0 // indirect
diff --git a/tests/fixtures/lock_generate_go/go.sum b/tests/fixtures/lock_generate_go/go.sum
index b0fa8e1..91a4d60 100644
--- a/tests/fixtures/lock_generate_go/go.sum
+++ b/tests/fixtures/lock_generate_go/go.sum
@@ -1,3 +1,4 @@
+github.com/coreos/go-systemd/v22 v22.5.0 h1:RrqgGjYQKalulkV8NGVIfkXQf6YYmOyiJKk8iXXhfZs=
 github.com/coreos/go-systemd/v22 v22.5.0/go.mod h1:Y58oyj3AT4RCenI/lSvhwexgC+NSVTIJ3seZv2GcEnc=
 github.com/go-chi/chi/v5 v5.0.12 h1:9euLV5sTrTNTRUU9POmDUvfxyj6LAABLUcEWO+JJb4s=
 github.com/go-chi/chi/v5 v5.0.12/go.mod h1:DslCQbL2OYiznFReuXYUmQ2hGd1aDpCnlMNITLSKoi8=
```

So once code is actually required, it adds the hash to `go.sum`. But more importantly, **it adds the dependency to `go.mod`**!!

### What do the docs say?

The [docs][1] for `go.mod` state:

[1]: https://go.dev/doc/modules/gomod-ref#go-notes

> - At go 1.17 or higher:
>   - The go.mod file includes an explicit require directive for each module that provides any package transitively imported by a package or test in the main module. (At go 1.16 and lower, an indirect dependency is included only if minimal version selection would otherwise select a different version.) This extra information enables module graph pruning and lazy module loading.

In other words, for **all supported versions of go** (currently oldest supported is 1.21), `go.mod` includes all direct and transitive dependencies with the precise version that will be used.

### Conclusion (TLDR)

We have been going about this all wrong... `go.mod` is a lockfile and we do not need `go.sum` at all.

**Review by kylewillmon — CHANGES_REQUESTED** [review]
It seems we have completely misunderstood the Go ecosystem... `go.mod` is a lockfile and we should parse it rather than going through all of this trouble to generate something equivalent.

File: tests/fixtures/lock_generate_go/go.sum (hunk 2 above)
Comment (kylewillmon) [reply tests/fixtures/lock_generate_go/go.sum:1]: > In other words, for **all supported versions of go** (currently oldest supported is 1.21), `go.mod` includes all direct and transitive dependencies with the precise version that will be used.

This is unfortunately not true... The Go version that is relevant is the `go` directive in `go.mod`, which is the minimum supported go version and is allowed to be lower than the oldest support go version... That complicates things slightly and puts us into a world similar to Python where `go.mod` **might be** a lockfile, but we don't know until we attempt to parse it.

Comment (ejortega) [reply tests/fixtures/lock_generate_go/go.sum:1]: For that same example, if you run `go list -m -json all`. You'll get the following:
```json
root@8d18781e51ee:/app# go list -m -json all
{
        "Path": "cli/example",
        "Main": true,
        "Dir": "/app",
        "GoMod": "/app/go.mod",
        "GoVersion": "1.22.2"
}
{
        "Path": "github.com/coreos/go-systemd/v22",
        "Version": "v22.5.0",
        "Time": "2022-11-07T13:52:27Z",
        "Indirect": true,
        "Dir": "/go/pkg/mod/github.com/coreos/go-systemd/v22@v22.5.0",
        "GoMod": "/go/pkg/mod/cache/download/github.com/coreos/go-systemd/v22/@v/v22.5.0.mod",
        "GoVersion": "1.12"
}
{
        "Path": "github.com/godbus/dbus/v5",
        "Version": "v5.0.4",
        "Time": "2021-03-23T03:10:11Z",
        "Indirect": true,
        "GoMod": "/go/pkg/mod/cache/download/github.com/godbus/dbus/v5/@v/v5.0.4.mod",
        "GoVersion": "1.12"
}
{
        "Path": "github.com/mattn/go-colorable",
        "Version": "v0.1.13",
        "Time": "2022-08-15T05:53:26Z",
        "Indirect": true,
        "Dir": "/go/pkg/mod/github.com/mattn/go-colorable@v0.1.13",
        "GoMod": "/go/pkg/mod/cache/download/github.com/mattn/go-colorable/@v/v0.1.13.mod",
        "GoVersion": "1.15"
}
{
        "Path": "github.com/mattn/go-isatty",
        "Version": "v0.0.19",
        "Time": "2023-03-22T15:51:17Z",
        "Indirect": true,
        "Dir": "/go/pkg/mod/github.com/mattn/go-isatty@v0.0.19",
        "GoMod": "/go/pkg/mod/cache/download/github.com/mattn/go-isatty/@v/v0.0.19.mod",
        "GoVersion": "1.15"
}
{
        "Path": "github.com/pkg/errors",
        "Version": "v0.9.1",
        "Time": "2020-01-14T19:47:44Z",
        "Indirect": true,
        "GoMod": "/go/pkg/mod/cache/download/github.com/pkg/errors/@v/v0.9.1.mod"
}
{
        "Path": "github.com/rs/xid",
        "Version": "v1.5.0",
        "Time": "2023-04-12T03:56:21Z",
        "Indirect": true,
        "GoMod": "/go/pkg/mod/cache/download/github.com/rs/xid/@v/v1.5.0.mod",
        "GoVersion": "1.12"
}
{
        "Path": "github.com/rs/zerolog",
        "Version": "v1.32.0",
        "Time": "2024-02-01T00:56:59Z",
        "Dir": "/go/pkg/mod/github.com/rs/zerolog@v1.32.0",
        "GoMod": "/go/pkg/mod/cache/download/github.com/rs/zerolog/@v/v1.32.0.mod",
        "GoVersion": "1.15"
}
{
        "Path": "golang.org/x/sys",
        "Version": "v0.12.0",
        "Time": "2023-08-25T20:53:44Z",
        "Indirect": true,
        "Dir": "/go/pkg/mod/golang.org/x/sys@v0.12.0",
        "GoMod": "/go/pkg/mod/cache/download/golang.org/x/sys/@v/v0.12.0.mod",
        "GoVersion": "1.17"
}
```

There's a few more dependencies that are not listed in the current `go.mod` file. I'm guessing as those functions/modules are not used in the current example or maybe because the mod file itself is `go 1.15`?

`github.com/pkg/errors` for example is listed as a dependency in the `go list` results because it's a direct dependency of `github.com/rs/zerolog`,  but not showing as an indirect in the current go.mod.

If you look at the mod file for `github.com/rs/zerolog`, you'll see `github.com/pkg/errors` is a direct dependency.
```bash
root@8d18781e51ee:/app# cat /go/pkg/mod/github.com/rs/zerolog\@v1.32.0/go.mod
module github.com/rs/zerolog

go 1.15

require (
        github.com/coreos/go-systemd/v22 v22.5.0
        github.com/mattn/go-colorable v0.1.13
        github.com/mattn/go-isatty v0.0.19 // indirect
        github.com/pkg/errors v0.9.1
        github.com/rs/xid v1.5.0
        golang.org/x/sys v0.12.0 // indirect
)
```

Comment (kylewillmon) [reply tests/fixtures/lock_generate_go/go.sum:1]: > `github.com/pkg/errors` for example is listed as a dependency in the `go list` results because it's a direct dependency of `github.com/rs/zerolog`, but not showing as an indirect in the current go.mod.

This is due to module pruning. `github.com/pkg/errors` is unused, so it is not included in the build and cannot effect the runtime behavior. For that same reason, we do not need to include it in Phylum analysis.

**Comment by ejortega** [comment]
Closing this PR as we'll be parsing the `go.mod` and `go.sum` directly.
